doc/developer/design/20230421_stabilize_monotonic_select.md

```diff
@@ -102,21 +102,21 @@ There are two strategies that we propose to lower the expense of monotonic opera
 
 The logic in the analysis above is similar to what is implemented at MIR level in the `MonotonicFlag` analysis. However, we consider only single-time contexts and a different level of abstraction, since the analysis is performed in LIR.
 
-The full set of single-time monotonicity enforcers is: `Plan::Reduce`, `Plan::TopK`, `Plan::Threshold`, `Plan::ArrangeBy`, `Plan::Constant` iff the constant is an `EvalError` or all its rows have `Diff` values greater than zero, and `Plan::Get` iff one of the following conditions is true: (1) the correspoding `GetPlan` is either `GetPlan::PassArrangements` or `GetPlan::Arrangement`; (2) The corresponding `GetPlan` is `GetPlan::Collection`, the `id` in `Plan::Get` is a local or global ID for a monotonic source or for an index.
+The full set of single-time monotonicity enforcers is: `Plan::Reduce`, `Plan::TopK`, `Plan::Threshold`, `Plan::Constant` iff the constant is an `EvalError` or all its rows have `Diff` values greater than zero, and `Plan::Get` iff the `id` in `Plan::Get` is a local or global ID for a non-recursive and monotonic input. For the latter, special care must be taken to keep a binding map with partial analysis results as `Plan::Let` nodes are visited.
 
-For monotonicity breakers, we note that any `Plan` nodes contained in the `values` of a `Plan::LetRec` expression are not in a single-time context due to iteration. This is why `refine_single_time_operator_selection` only considers the `body` of a `Plan::LetRec`, explicitly excluding traversal of `values`. So we must be careful in the analysis to consider the bindings introduced by `Plan::LetRec` to be monotonicity breakers. Additionally, monotonicity breakers comprise `Plan::Negate`, `Plan::Constant` iff the constant contains rows with non-positive `Diff` values, and `Plan::Get` nodes iff the corresponding `GetPlan` is `GetPlan::Collection` and the `id` in `Plan::Get` is either a local or global ID for a non-monotonic source.
+For monotonicity breakers, we note as remarked above that any `Plan` nodes contained in the `values` of a `Plan::LetRec` expression are not in a single-time context due to iteration. This is why `refine_single_time_operator_selection` only considers the `body` of a `Plan::LetRec`, explicitly excluding traversal of `values`. So we must be careful in the analysis to consider the bindings introduced by `Plan::LetRec` to be monotonicity breakers. Additionally, monotonicity breakers comprise `Plan::Negate`, `Plan::Constant` iff the constant contains rows with non-positive `Diff` values, and `Plan::Get` nodes if the `id` in `Plan::Get` is either a local or global ID for a non-monotonic input.
```

Comment:
For `Plan::LetRec`, I think we can improve this in a future PR so "non-recursive" paths within a `LetRec` block will be treated as they were not in a `LetRec` block. For example, in

```sql
WITH MUTUALLY RECURSIVE
  c_0 AS $expr_0[c_2],
  c_1 AS $expr_1[not c_0, not c_2],
  c_2 AS $expr_2[c_2]
$return[c_0, c_1, c_2]
```

the definition of the second binding (`c_1`) does not depend on `c_0` or on `c_2`, so it will be executed only once, even though it is rendered in a recursive context. Does this make sense?

Reply:
Nice, very interesting observation. Should this be part of a future PR refining the analysis or could it be achieved by an earlier transform to hoist these definitions out into a vanilla `Let`?